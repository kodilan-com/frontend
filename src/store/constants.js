export const TOGGLE_LOADING = 'TOGGLE_LOADING';
export const SET_RECENT_POSTS = 'SET_RECENT_POSTS';
export const SET_ALL_POSTS = 'SET_ALL_POSTS';
export const SET_FEATURED = 'SET_FEATURED';
export const SET_TAGS = 'SET_TAGS';
export const SET_AVAILABLE_LOCATIONS = 'SET_AVAILABLE_LOCATIONS';
export const SET_ACTIVE_PERIOD = 'SET_ACTIVE_PERIOD';
export const RECENT_POST_COUNT = 25;
export const JOB_TYPE_MAP = {
  0: 'Tam zamanlı',
  1: 'Tam zamanlı',
  2: 'Yarı zamanlı',
  3: 'Stajyer',
  4: 'Freelance',
};

export const PERIODS = [
  {
    type: 'daily',
    text: 'Bugün',
    slug: 'bugun',
  },
  {
    type: 'weekly',
    text: 'Bu hafta',
    slug: 'bu-hafta',
  },
  {
    type: 'monthly',
    text: 'Bu ay',
    slug: 'bu-ay',
    isDefault: true,
  },
];

export const JOB_TYPE_SLUG_TO_ID_MAP = {
  'tam-zamanli': 1,
  'yari-zamanli': 2,
  stajyer: 3,
  freelance: 4,
};

export const JOB_TYPE_ID_TO_SLUG_MAP = {
  1: 'tam-zamanli',
  2: 'yari-zamanli',
  3: 'stajyer',
  4: 'freelance',
};

export const JOB_TYPES_FOR_DROPDOWN = [
  {
    id: 1,
    text: 'Tam zamanlı',
  },
  {
    id: 2,
    text: 'Yarı zamanlı',
  },
  {
    id: 3,
    text: 'Stajyer',
  },
  {
    id: 4,
    text: 'Freelance',
  },
];
