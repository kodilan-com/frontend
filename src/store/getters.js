export default {
  autocompleteTags(state) {
    return state.tags.map(tag => ({ label: tag.name, value: tag.slug }));
  },
  companyId(state) {
    return Object.keys(state.me).length && state.me.companies[0].id;
  },
};
