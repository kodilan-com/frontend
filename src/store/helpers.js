export default {
  rankPosts(post, postTagSlugs, relatedPosts) {
    const DAY_IN_MS = 1000 * 60 * 60 * 24;
    const TODAY = new Date();
    const POINTS = {
      createdInTenDays: 2,
      fromSameCompany: 3,
      hasSamePostType: 4,
      hasSameTag: 5,
      inSameLocation: 6,
    };

    return relatedPosts
      .filter(p => p.slug !== post.slug)
      .map((p) => {
        let score = 0;

        score += p.location === post.location ? POINTS.inSameLocation : 0;
        score += p.company.slug === post.company.slug ? POINTS.fromSameCompany : 0;
        score += p.type === post.type ? POINTS.hasSamePostType : 0;
        score += (TODAY - new Date(p.updated_at)) / DAY_IN_MS <= 10 ? POINTS.createdInTenDays : 0;
        p.tags.forEach((t) => {
          score += postTagSlugs.indexOf(t.slug) > -1 ? POINTS.hasSameTag : 0;
        });

        return { score, post: p };
      })
      .sort((c, n) => (c.score > n.score ? -1 : 1))
      .slice(0, 3)
      .map(o => o.post);
  },
};
