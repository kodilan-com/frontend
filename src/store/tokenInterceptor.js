export default (acc) => {
  const token = localStorage.getItem('AccessToken');
  if (token !== null) {
    acc.headers.Authorization = `Bearer ${token}`;
  }
  return acc;
};
