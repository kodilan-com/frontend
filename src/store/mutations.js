import * as constants from './constants';

export default {
  [constants.TOGGLE_LOADING](state) {
    state.isLoading = !state.isLoading;
  },
  [constants.SET_RECENT_POSTS](state, recentPosts) {
    state.recentPosts = recentPosts;
  },
  [constants.SET_ALL_POSTS](state, allPosts) {
    state.allPosts = allPosts;
  },
  [constants.SET_ALL_POST_META](state, meta) {
    state.allPostsMeta = meta;
  },
  [constants.SET_FEATURED](state, featuredPosts) {
    state.featuredPosts = featuredPosts;
  },
  [constants.SET_TAGS](state, tags) {
    state.tags = tags;
  },
  [constants.SET_AVAILABLE_LOCATIONS](state, locations) {
    state.availableLocations = locations.reduce((arr, loc) => {
      if (loc.location !== 'Remote') {
        arr.push(loc.location);
      }

      return arr;
    }, ['Remote']);
  },
  [constants.SET_ACTIVE_PERIOD](state, period) {
    state.activePeriod = period;
  },
  setIsLoggedIn(state, data) {
    state.isLoggedIn = data;
  },
  SET_ME(state, data) {
    state.me = data;
  },
  postListing(state, data) {
    state.manageJobs = data;
  },
};
