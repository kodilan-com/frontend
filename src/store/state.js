import { PERIODS } from './constants';

export default {
  isLoggedIn: localStorage.getItem('AccessToken') || null,
  me: {},
  manageJobs: [],
  isLoading: true,
  recentPosts: [],
  activePeriod: PERIODS.find(period => period.isDefault).type,
  featuredPosts: [],
  allPosts: [],
  allPostsMeta: {},
  tags: [],
  availableLocations: ['Remote', 'İstanbul', 'İzmir', 'Ankara', 'Bursa'],
};
