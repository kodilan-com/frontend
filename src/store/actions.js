import axios from 'axios';
import { cacheAdapterEnhancer } from 'axios-extensions';
import * as constants from './constants';
import helpers from './helpers';

const http = axios.create({
  baseURL: 'https://api.kodilan.com',
  adapter: cacheAdapterEnhancer(axios.defaults.adapter),
});

export default {
  toggleLoading({ commit }) {
    commit(constants.TOGGLE_LOADING);
  },
  fetchRecentPosts({ state, commit }) {
    return http.get(`/posts?get=${constants.RECENT_POST_COUNT}&period=${state.activePeriod}`)
      .then((res) => {
        commit(constants.SET_RECENT_POSTS, res.data.data);

        return res.data;
      });
  },
  fetchFeaturedPosts({ commit }) {
    return http.get('/posts?get=3&is_featured=1')
      .then((res) => {
        commit(constants.SET_FEATURED, res.data.data);

        return res.data;
      });
  },
  fetchAllPosts({ commit }, payload = {}) {
    const page = payload
    && payload.page
    && Number.isInteger(payload.page)
    && payload.page > 1 ? payload.page : 1;

    return http.get(`/posts?get=${constants.PER_PAGE}&page=${page}`)
      .then((res) => {
        const { data } = res;
        commit(constants.SET_ALL_POSTS, data.data);
        commit(constants.SET_ALL_POST_META, {
          total: data.total,
          current_page: data.current_page,
          last_page: data.last_page,
        });

        return res.data;
      });
  },
  fetchBySlug(_, slug) {
    return http.get(`/posts/${slug}`)
      .then(res => res.data);
  },
  fetchByCompany(_, company) {
    return http.get(`/companies/${company}/posts`)
      .then(res => res.data);
  },
  fetchByTag(_, tag) {
    return http.get(`/tags/${tag}/posts`)
      .then(res => res.data);
  },
  search(_, params) {
    return http.get('/search', { params })
      .then(res => res.data);
  },
  fetchTags({ commit }) {
    return http.get('/tags')
      .then((res) => {
        commit(constants.SET_TAGS, res.data.data);
      });
  },
  fetchRelatedPosts({ dispatch }, post) {
    const postTags = post.tags.map(t => t.slug);
    const categoryTags = ['frontend', 'backend', 'mobile', 'designer', 'qa', 'devops'];
    const [mainTag] = postTags.filter(t => categoryTags.indexOf(t) > -1);

    if (!mainTag) {
      return Promise.resolve([]);
    }

    return dispatch('fetchByTag', mainTag)
      .then(res => helpers.rankPosts(post, postTags, res.data));
  },
  fetchAvailableLocations({ commit }) {
    return http.get('/posts/locations')
      .then((res) => {
        commit(constants.SET_AVAILABLE_LOCATIONS, res.data);
      });
  },
  savePost(_, data) {
    return http.post('/posts', data);
  },
  subscribe(_, data) {
    return http.post('/newsletters', data);
  },
  setPeriod({ commit }, period) {
    commit('SET_ACTIVE_PERIOD', period);
  },
};
