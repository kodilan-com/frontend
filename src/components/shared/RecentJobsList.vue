<script>
import { mapState, mapActions } from 'vuex';
import JobListing from './JobListing';
import JobPeriodSelector from './JobPeriodSelector';
import { PERIODS } from '../../store/constants';

export default {
  components: {
    JobListing,
    JobPeriodSelector,
  },
  data() {
    return {
      isLoading: false,
    };
  },
  computed: {
    ...mapState(['recentPosts', 'activePeriod']),
  },
  methods: {
    ...mapActions(['fetchRecentPosts', 'setPeriod']),
    fetch() {
      this.isLoading = true;

      this.fetchRecentPosts()
        .then(() => {
          this.isLoading = false;
        });
    },
    setPeriodHandler() {
      const period = PERIODS.find(p => p.type === this.$route.meta.period);
      if (period) {
        this.setPeriod(period.type);
      }
      this.fetch();
    },
  },
  watch: {
    $route(to, from) {
      if (to.meta.period !== from.meta.period) this.setPeriodHandler();
    },
  },
  created() {
    this.setPeriodHandler();
  },
};
</script>

<template>
  <div class="eleven columns">
    <div class="padding-right">
      <h3>
        <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor">
          <path stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6 5c7.18 0 13 5.82 13 13M6 11a7 7 0 017 7m-6 0a1 1 0 11-2 0 1 1 0 012 0z" />
        </svg>
        En son eklenen ilanlar
      </h3>
      <div class="margin-bottom-30" />
      <job-period-selector />
      <div class="margin-bottom-25" />
      <job-listing
        :is-loading="isLoading"
        :posts="recentPosts"
      />
      <router-link
        to="/ilanlar"
        class="button centered tag-all-posts"
        v-if="!isLoading && recentPosts.length"
      >
        <!--<i class="fa fa-plus-circle" />-->
        <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor">
          <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M12 9v3m0 0v3m0-3h3m-3 0H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z" />
        </svg>
        Tüm ilanları gör
      </router-link>
      <div class="margin-bottom-55" />
    </div>
  </div>
</template>

<style lang="scss" scoped>
  h3 {
    display: flex;
    flex-direction: row;
    align-items: center;

    font-weight: 500;

    svg {
      width: 24px;
      height: 24px;
      background-color: #26ae61;
      stroke: #FFF;
      margin-right: 12px;
      padding: 8px;
      border-radius: 50%;
    }
  }

</style>
