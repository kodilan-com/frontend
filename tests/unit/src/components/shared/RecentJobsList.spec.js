import RecentJobsList from '@/components/shared/RecentJobsList';
import { createLocalVue, shallowMount, RouterLinkStub } from '@vue/test-utils';
import Vuex from 'vuex';
import { PERIODS } from '@/store/constants';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('RecentJobsList.vue', () => {
  let wrapper;
  let storeStub;
  let store;
  const setPeriodHandlerStub = jest.fn();
  let $route;
  const renderWrapper = (props, methods) => {
    wrapper = shallowMount(RecentJobsList, {
      localVue,
      store,
      mocks: {
        $route,
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
      propsData: {
        ...props,
      },
      methods: {
        ...methods,
      },
    });
  };

  beforeEach(() => {
    $route = {
      meta: {
        period: '',
      },
    };
    storeStub = {
      state: {
        recentPosts: ['test1', 'test2'],
        activePeriod: 'testPeriod',
      },
      actions: {
        fetchRecentPosts: jest.fn(),
        setPeriod: jest.fn(),
      },
    };

    store = new Vuex.Store(storeStub);

    renderWrapper();
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();

    wrapper.destroy();
  });

  it('should match with snapshot', () => {
    renderWrapper({}, { setPeriodHandler: setPeriodHandlerStub });

    expect(wrapper.element)
      .toMatchSnapshot();
  });

  it('should call setPeriodHandler method on initial', async () => {
    renderWrapper({}, { setPeriodHandler: setPeriodHandlerStub });

    await localVue.nextTick();

    expect(setPeriodHandlerStub)
      .toBeCalled();
  });

  it('should set "/ilanlar" to link', () => {
    renderWrapper({}, { setPeriodHandler: setPeriodHandlerStub });

    expect(wrapper.find('.tag-all-posts')
      .props('to'))
      .toBe('/ilanlar');
  });

  it('should not render link if isLoading is true', async () => {
    renderWrapper({}, { setPeriodHandler: setPeriodHandlerStub });

    wrapper.setData({
      isLoading: true,
    });

    await localVue.nextTick();

    expect(wrapper.find('.tag-all-posts')
      .exists())
      .toBe(false);
  });

  it('should call fetch method and set period if route period exists in PERIODS', () => {
    const fetchStub = jest.fn();
    const setPeriodStub = jest.fn();

    $route.meta.period = PERIODS[1].type;

    renderWrapper({}, {
      setPeriod: setPeriodStub,
      fetch: fetchStub,
    });

    wrapper.vm.setPeriodHandler();

    expect(fetchStub)
      .toBeCalled();
    expect(setPeriodStub)
      .toBeCalledWith(PERIODS[1].type);
  });

  it('should call fetch method and not set period if route period not exists in PERIODS', () => {
    const fetchStub = jest.fn();
    const setPeriodStub = jest.fn();

    $route.meta.period = 'notExists';

    renderWrapper({}, { setPeriod: setPeriodStub, fetch: fetchStub });

    wrapper.vm.setPeriodHandler();

    expect(fetchStub).toBeCalled();
    expect(setPeriodStub).not.toBeCalled();
  });
});
