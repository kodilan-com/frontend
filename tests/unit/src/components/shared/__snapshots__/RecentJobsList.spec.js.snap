// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentJobsList.vue should match with snapshot 1`] = `
<div
  class="eleven columns"
>
  <div
    class="padding-right"
  >
    <h3
      class="margin-bottom-25"
    >
      
      En son eklenen ilanlar
    
    </h3>
     
    <job-period-selector-stub />
     
    <job-listing-stub
      notfoundtext="İlan bulunamadı."
      posts="test1,test2"
    />
     
    <a
      class="button centered tag-all-posts"
    >
      <i
        class="fa fa-plus-circle"
      />
      
      Tüm ilanları gör
    
    </a>
     
    <div
      class="margin-bottom-55"
    />
  </div>
</div>
`;
